### PR TITLE
Reject queries and commands with duplicated IDs

### PR DESCRIPTION
**Pull request by trimoq**
All feedback welcome, especially if someone has a better name for the new interface.

### CODE REVIEW

File: axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandCache.java [hunk 1]
```diff
@@ -105,6 +105,15 @@ public CommandInformation put(@Nonnull String key, @Nonnull CommandInformation v
         return map.put(key, value);
     }
 
+    /**
```
Comment (MGathier) [inline axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandCache.java:108]: Do you still want to allow the normal put() operation here?

Comment (MGathier) [reply axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandCache.java:108]: Same applies for the QueryCache

**Review by MGathier — COMMENTED** [review]
We want to include the changes in a 4.6.x patch, so make the PR for the 4.6.x branch instead of master.

File: axonserver/src/main/java/io/axoniq/axonserver/util/NonReplacingConstraintCache.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package io.axoniq.axonserver.util;
+
+public interface NonReplacingConstraintCache <K,V> extends ConstraintCache<K,V>{
```
Comment (m1l4n54v1c) [inline axonserver/src/main/java/io/axoniq/axonserver/util/NonReplacingConstraintCache.java:3]: javadoc.

File: axonserver/src/test/java/io/axoniq/axonserver/message/command/CommandDispatcherTest.java [hunk 3]
```diff
@@ -322,4 +326,48 @@ public SerializedCommandResponse commandResponse(
...
+
+        commandDispatcher.handleResponse(new SerializedCommandResponse(CommandResponse.newBuilder()
+                                                                                      .setRequestIdentifier(duplicatedId)
+                                                                                      .build()), false);
+
+        try {
+            SerializedCommandResponse originalResponse = originalFutureResponse.get(1, TimeUnit.SECONDS);
+            assertEquals(duplicatedId, originalResponse.getRequestIdentifier());
+        } catch (TimeoutException e) {
+            fail();
```
Comment (m1l4n54v1c) [inline axonserver/src/test/java/io/axoniq/axonserver/message/command/CommandDispatcherTest.java:367]: I don't believe there is need for this: if the `TimeoutException` is thrown, the test will fail anyway :)

File: axonserver/src/test/java/io/axoniq/axonserver/message/query/QueryDispatcherTest.java [hunk 4]
```diff
@@ -439,6 +437,74 @@ public void queryFlowControl() throws InterruptedException {
...
+
+        testSubject.handleComplete(requestId, "client", "clientId", false);
+
+        // would not finish, if AS would override query in query cache
+        try {
+            QueryResponse originalResponse = originalFutureResponse.get(1, TimeUnit.SECONDS);
+            assertTrue(originalFutureCompleted.get(1, TimeUnit.SECONDS));
+            assertEquals(requestId, originalResponse.getRequestIdentifier());
+        } catch (TimeoutException e) {
+            fail();
```
Comment (m1l4n54v1c) [inline axonserver/src/test/java/io/axoniq/axonserver/message/query/QueryDispatcherTest.java:500]: Unnecessary.

**Review by m1l4n54v1c — APPROVED** [review]
Looks good, minor comments.

File: axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandCache.java (hunk 1 above)
Comment (trimoq) [reply axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandCache.java:102]: Removed in the new commit: `put` is deprecated in the child-interface and an exception throwing default implementation is provided.

**Review by MGathier — APPROVED** [review]

**Pull request merged**